pkg/sources/gitlab/gitlab.go

```diff
@@ -262,119 +263,89 @@ func (s *Source) getRepos() ([]*url.URL, []error) {
 
 func (s *Source) scanRepos(ctx context.Context, chunksChan chan *sources.Chunk, repos []*url.URL) []error {
 	wg := sync.WaitGroup{}
-	errs := []error{}
+	var errs []error
 	var errsMut sync.Mutex
-	if s.authMethod == "UNAUTHENTICATED" {
-		for i, u := range repos {
-			if common.IsDone(ctx) {
-				// We are returning nil instead of the scanErrors slice here because
-				// we don't want to mark this scan as errored if we cancelled it.
-				return nil
+	errs = s.scan(ctx, chunksChan, repos, &wg, &errsMut, errs)
+	wg.Wait()
+
+	return errs
+}
+
+func (s *Source) scan(ctx context.Context, chunksChan chan *sources.Chunk, repos []*url.URL, wg *sync.WaitGroup, errsMut *sync.Mutex, errs []error) []error {
+	for i, u := range repos {
+		if common.IsDone(ctx) {
+			// We are returning nil instead of the scanErrors slice here because
+			// we don't want to mark this scan as errored if we cancelled it.
+			return nil
+		}
+		if err := s.jobSem.Acquire(ctx, 1); err != nil {
+			log.WithError(err).Debug("could not acquire semaphore")
+			continue
+		}
+		wg.Add(1)
+		go func(ctx context.Context, repoURL *url.URL, i int) {
+			defer s.jobSem.Release(1)
+			defer wg.Done()
+			if len(repoURL.String()) == 0 {
+				return
 			}
-			if err := s.jobSem.Acquire(ctx, 1); err != nil {
-				log.WithError(err).Debug("could not acquire semaphore")
-				continue
+			s.SetProgressComplete(i, len(repos), fmt.Sprintf("Repo: %s", repoURL), "")
+
+			var path string
+			var repo *git2.Repository
+			var err error
+			if s.authMethod == "UNAUTHENTICATED" {
+				path, repo, err = git.CloneRepoUsingUnauthenticated(repoURL.String())
+			} else {
+				path, repo, err = git.CloneRepoUsingToken(s.token, repoURL.String(), s.user)
 			}
-			wg.Add(1)
-			go func(ctx context.Context, repoURL *url.URL, i int) {
-				defer s.jobSem.Release(1)
-				defer wg.Done()
-				if len(repoURL.String()) == 0 {
-					return
-				}
-				s.SetProgressComplete(i, len(repos), fmt.Sprintf("Repo: %s", repoURL), "")
```

Comment:
These calls are needed to update progress

Reply:
Oh yea i don't think i removed any of them. It's just moved up to line 292. I simply refactored both the conditionals into a single since the only difference i noticed was `git. CloneRepoUsingUnauthenticated` vs `git. CloneRepoUsingToken` the rest should be the same.